homeassistant/components/starlink/__init__.py

```diff
@@ -0,0 +1,155 @@
+"""The Starlink integration."""
+from __future__ import annotations
+
+from abc import ABC
+from datetime import timedelta
+import logging
+
+import async_timeout
+from spacex.starlink import CommunicationError, StarlinkDish
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ADDRESS, CONF_NAME, CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA_BASE
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    CONF_DEFAULT_ADDRESS,
+    CONF_DEFAULT_NAME,
+    CONF_DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[str] = ["binary_sensor", "sensor"]
+PLATFORM_SCHEMA_BASE.extend(
+    {
+        vol.Optional(CONF_NAME, default=CONF_DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_ADDRESS, default=CONF_DEFAULT_ADDRESS): cv.string,
+        vol.Optional(
+            CONF_SCAN_INTERVAL, default=CONF_DEFAULT_SCAN_INTERVAL
+        ): cv.positive_float,
+    }
+)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Starlink from a config entry."""
+    dish = StarlinkDish(address=entry.data[CONF_ADDRESS], autoconnect=True)
+
+    async def async_update_data():
+        async with async_timeout.timeout(10):
+            if not dish.connected:
+                try:
+                    dish.connect(refresh=False)  # Avoid refreshing twice
+                    _LOGGER.info("Connected to the satellite")
+                except CommunicationError as err:
+                    raise UpdateFailed(f"Could not connect to satellite: {err}")
+            try:
+                dish.refresh()
+            except CommunicationError as err:
+                _LOGGER.warning("Lost connection to Dishy")
+                dish.close()
+                raise UpdateFailed(f"Error communicating with API: {err}")
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=entry.data[CONF_NAME],
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=entry.data[CONF_SCAN_INTERVAL] or 5),
+    )
+
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][entry.entry_id] = (dish, coordinator)
+
+    await coordinator.async_refresh()
```

Comment:
`await coordinator.async_config_entry_first_refresh()` could be used instead and should then be put above L71